tests/kernel/userbufer_validate/src/userbuffer_validate.c

```diff
@@ -0,0 +1,381 @@
+ * else return 0.
+ * Below conditions will be tested accordingly
+ *
+ */
+
+static int buffer_rw_read(void)
+{
+	/* read write testing */
+	x86_mmu_set_pte_rw_bit((void *)0x00405000, 0);
+	status = _x86_mmu_buffer_validate((void *)0x00405000,
+					   10, BUFF_WRITEABLE);
+	/* should fail */
+	if (status != -EPERM) {
+		TC_PRINT("buffer_rw_read failed\n");
+		return TC_FAIL;
+	}
+
+	return TC_PASS;
+}
+
+static int buffer_writeable_write(void)
+{
+	x86_mmu_set_pte_rw_bit((void *)0x00405000, 1);
+	status = _x86_mmu_buffer_validate((void *)0x00405000, 10, 0x1);
+	/* should pass */
+	if (status != 0) {
+		TC_PRINT("buffer_rw_write failed\n");
+		return TC_FAIL;
+	}
+
+	return TC_PASS;
+}
+
+static int buffer_readable_read(void)
+{
+	x86_mmu_set_pte_rw_bit((void *)0x00405000, 0);
+	status = _x86_mmu_buffer_validate((void *)0x00405000, 10, 0x0);
+	/* should pass */
+	if (status != 0) {
+		TC_PRINT("buffer_readble_read failed\n");
+		return TC_FAIL;
+	}
+
+	return TC_PASS;
+}
+
+static int buffer_readable_write(void)
+{
+	x86_mmu_set_pte_rw_bit((void *)0x00405000, 1);
+	status = _x86_mmu_buffer_validate((void *)0x00405000, 10, 0x0);
+	/* should pass */
+	if (status != 0) {
+		TC_PRINT("buffer_readable_write failed\n");
+		return TC_FAIL;
+	}
+
+	return TC_PASS;
+}
+
+static int buffer_supervisor_rw(void)
+{
+	/* user/supervisor test */
+	x86_mmu_set_pte_us_bit((void *)0x00405000, 0); /* supervisor */
+	status = _x86_mmu_buffer_validate((void *)0x00405000,
+					   10, 0x1 | 0x2);
+	/* should fail */
+	if (status != -EPERM) {
+		TC_PRINT("buffer_supervisor_rw failed\n");
+		return TC_FAIL;
+	}
+
+	return TC_PASS;
+}
+
+static int buffer_supervisor_w(void)
+{
+	x86_mmu_set_pte_us_bit((void *)0x00405000, 0); /* supervisor */
+	status = _x86_mmu_buffer_validate((void *)0x00405000,
+					   10, 0x1 | 0x0);
+	/* should pass */
+	if (status != 0) {
+		TC_PRINT("buffer_supervisor_w failed\n");
+		return TC_FAIL;
+	}
+
+	return TC_PASS;
+}
+
+static int buffer_user_rw_user(void)
+{
+	x86_mmu_set_pde_us_bit((void *)0x00405000, 1); /* set pde user */
+	x86_mmu_set_pte_us_bit((void *)0x00405000, 1); /* user */
+	status = _x86_mmu_buffer_validate((void *)0x00405000,
+					   10, 0x1 | 0x2);
+	/* should pass */
+	if (status != 0) {
+		TC_PRINT("buffer_user_rw_user failed\n");
+		return TC_FAIL;
+	}
+
+	return TC_PASS;
+}
+
+static int buffer_user_rw_supervisor(void)
+{
+	x86_mmu_set_pde_us_bit((void *)0x00405000, 1); /*set pde user*/
+	x86_mmu_set_pte_us_bit((void *)0x00405000, 0); /* supervisor */
+	status = _x86_mmu_buffer_validate((void *)0x00405000,
+					   10, 0x1 | 0x2);
+	/* should fail */
+	x86_mmu_set_pde_us_bit((void *)0x00405000, 0);  /* reset */
+	if (status != -EPERM) {
+		TC_PRINT("buffer_user_rw_supervisor failed\n");
+		return TC_FAIL;
+	}
+
+	return TC_PASS;
+}
+
+/* tc for multiple page buffers */
+/* read write testing */
+static int multi_page_buffer_writeable(void)
+{
+	x86_mmu_set_pte_rw_bit((void *)0x00405000, 0);
```

Comment:
why are these addresses hard-coded?
you need to programmatically select memory addresses to test.
there are linker variables defined to help you with this. see include/linker/linker-defs.h

Reply:
hard-coded addresses have been removed, also the linker variables have been used